performancecopilot/pcp
# pmlogger and friends

Changes committed to git@github.com:kmcdonell/pcp.git 20200329

Ken McDonell (12):
      src/pmlogger/src/pmlogger.c: change duplicate archive name avoidance strategy
      src/pmlogger/pmlogger_check.sh: fortify test for archive creation
      src/libpcp/src/auxserver.c: hide warning from __pmServerNotifySystemd
      qa/src/.gitignore: add test_service_notify
      qa/041: don't need to dodge systemd here
      qa/1347: the shell in !# is /bin/sh, not /usr/bin/sh
      qa/340: save original .../pmlogger/control in the same dir, not /tmp
      qa/1896: refactor after changes to pmlogger duplicate avoidance scheme
      pmlogger et al: drop 0 in strftime patterns
      src/pmlogger/pmlogger_check.sh: don't update the Latest folio
      pmlogger: refactor -NN archive name collision avoidance
      qa: assorted flow-on changes from recent pmlogger changes


Details ...

commit e4119f66d33c7eb04f4b4d18e3873b654f0c00fb
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Mon Mar 30 06:53:11 2020 +1100

    qa: assorted flow-on changes from recent pmlogger changes

commit 3c5cee2e10d7768b113c5dd2bc136fa3a0b2abff
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Mon Mar 30 06:50:19 2020 +1100

    pmlogger: refactor -NN archive name collision avoidance
    
    The pre-test with access() is insufficient.
    
    Need to move __pmLogCreate() inside the -NN loop and break when this
    succeeds.
    
    A bunch of knock-on effects of this, as the real archive name is not
    known until the __pmLogCreate() succeeds, and may be different to the
    archive name after the initial strftime() expansion.

commit f4887081796782673250424bc8ec39eee61e7383
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Mon Mar 30 06:48:24 2020 +1100

    src/pmlogger/pmlogger_check.sh: don't update the Latest folio
    
    This is now done by pmlogger.

commit 26941037f4ee3e55ab01cb61dda359e401c23a5d
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 14:04:51 2020 +1100

    pmlogger et al: drop 0 in strftime patterns
    
    The changed duplicate avoidance scheme means the argv[] munging has
    gone away, and with it the need for the expanded archive base name
    to be no longer than the pattern from the command line.
    
    This puts things back the way they were, with the default archive
    basename pattern (for the system managed pmloggers) being:
            %Y%m%d.%H.%M

commit 0fa97ef37cdd714cf614efa367a281affcf399e0
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 14:02:28 2020 +1100

    qa/1896: refactor after changes to pmlogger duplicate avoidance scheme
    
    Don't need sleep and sync to avoid minute click over.
    
    Archive names are not deterministic, but only a small finite set of things
    need be tested for.

commit a95eec534c60b75eb30b7b2d9ba12ad8fd3093bd
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:39:02 2020 +1100

    qa/340: save original .../pmlogger/control in the same dir, not /tmp
    
    if the test aborts, we at least have control.340 left in .../pmlogger
    to know what the contents used to be.

commit 17eb9a64666406b045db1910c5bf12d0cfba9905
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:38:20 2020 +1100

    qa/1347: the shell in !# is /bin/sh, not /usr/bin/sh

commit a926f850bd118843b2c00a0c38cc8d08a684f4fd
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:37:28 2020 +1100

    qa/041: don't need to dodge systemd here
    
    It is now safe to restart with _service.

commit 0c44e9e853c452d622f07e703a1dc34bec660f15
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:36:41 2020 +1100

    qa/src/.gitignore: add test_service_notify

commit 0b5896d45057f153e4f3d02b0f9449941f145db4
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:32:39 2020 +1100

    src/libpcp/src/auxserver.c: hide warning from __pmServerNotifySystemd
    
    pmlogger will unconditonally call __pmServerNotifySystemd() now,
    but if pmlogger is not started by systemd (e.g. pmlogger launched by
    record mode, stand alone logging or (especially) the QA suite) then
    we don't really want the warning:
        __pmServerNotifySystemd: NOTIFY_SOCKET not set, not launched by systemd
    as this is expected behaviour.
    
    Move the warning inside a pmDebugOptions.services guard.

commit 83ee123e60a98a3b397c19963aaa32ed93a521f0
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:29:44 2020 +1100

    src/pmlogger/pmlogger_check.sh: fortify test for archive creation
    
    Previously we knew the base name of the newly created archive, but
    recent changes have meant that this decision is made by pmlogger and
    so this script is left to deduce the base name.
    
    If pmlogger does not start (like on purpose in some QA tests), then
    the archive is not created and the logic here was a bit busted.

commit ac5f94760223cfbc4d9f8aaf998a1972d951e273
Author: Ken McDonell <kenj@kenj.id.au>
Date:   Sun Mar 29 13:22:08 2020 +1100

    src/pmlogger/src/pmlogger.c: change duplicate archive name avoidance strategy
    
    When pmlogger is launched repeatedly from systemd (or anywhere using
    the strftime pattern for the archive name), there is a chance that more
    than one invocation happens inside a minute, in which case the %0M at
    the end of the strftime pattern is not enough to guarantee uniqueness.
    
    Previously we we just pretending that we were starting +1 or +2 or
    +3 miniutes later, but this approach is doomed in the presence of
    rapid restarting, like in QA.
    
    This commit reinstates the scheme from pmnewlog(1) (now deprecated)
    and uses suffixes like "", "-00", "-01", ... to force the output
    archive base name to be unique.
    
    In the process this means the real archive name maybe longer
    than the strftime pattern, so drop the argv[] munging, which on
    reflection appears to be (a) non-portable, (b) not reliable, and (c)
    not necessary.
